[JENKINS-46175] SimpleBuildStep/SimpleBuildWrapper without workspace

Zastai opened the pull request:
See [JENKINS-46175](https://issues.jenkins-ci.org/browse/JENKINS-46175).

This makes the workspace and launcher parameters nullable for
`SimpleBuildStep.perform()` (both overloads).

New `requiresLauncher()` and `requiresWorkspace()` predicates determine
whether they are really required; by default, these return `true`,
preserving the old non-null behaviour.

Similar changes were made for `SimpleBuildWrapper`, with the predicates
controlling the nullability of workspace/launcher on the `setUp` method.

Currently no tests added; this does not really change any behaviour, it only
enables plugins not to require a workspace/launcher context. It is `CoreStep`
and `CoreWrapperStep` in `workspace-basic-steps` that will get behaviour
changes. Setting this as draft until I have a PR ready on that plugin (using
an incremental from this one).

Note: initially I just had a single `requiresWorkspace()` for both; but then I saw
that `BuildStepCompatibilityLayer` seemed to support a case where workspace
and launcher were not present in lockstep, I added a separate predicate. This
seems harmless and allows more granular specification of needs.

### Proposed changelog entries

* JENKINS-46175: A `SimpleBuildStep` or `SimpleBuildWrapper` can now choose not to require a workspace and/or launcher.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->
Reference implementation:

* https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/117

### Desired reviewers

@mention jglick

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

Zastai committed:
[JENKINS-46175] SimpleBuildSteps without workspace

This makes the workspace and launcher parameters check-for-null for
SimpleBuildStep.perform() (both overloads).

New requiresLauncher() and requiresWorkspace() predicates determines
whether they are really required; by default, these returns true,
preserving the old non-null behaviour.

In BuildStepCompatibilityLayer, don't abort when there is no workspace unless
the SimpleBuildStep requires one.
Similarly, check the unannotated launcher parameter for null when the build
step requires a launcher.

Zastai committed:
[JENKINS-46175] SimpleBuildSteps without workspace

Fingerprinter does not use the launcher passed to perform(),
so add an overload of requiresLauncher() to make that explicit.

Zastai committed:
[JENKINS-46175] SimpleBuildWrappers without workspace

This adds predicates to SimpleBuildWrapper and
SimpleBuildWrapper.Disposer, to indicate whether a workspace and/or
launcher are required. As with SimpleBuildStep these all default
to true, to preserve the old behaviour by default.

This also adds explicit nullability annotations to all affected
methods, assuming non-null for the context, build and listener.

Zastai commented:
The behaviour is tested as part of https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/117, where it's `CoreStep` that will allow running `SimpleBuildStep`s outside of node context (and same for `CoreWrapper`).
See https://github.com/Zastai/workflow-basic-steps-plugin/blob/JENKINS-46175/src/test/java/org/jenkinsci/plugins/workflow/steps/CoreStepTest.java#L172-L220 and https://github.com/Zastai/workflow-basic-steps-plugin/blob/JENKINS-46175/src/test/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStepTest.java#L293-L363

I'm not entirely sure I could test this in Jenkins core; I'd have to see if there's anything in the test harness that allows me to run something other than `AbstractProject`, in order to call a step without passing workspace/launcher. But then I'm not familiar with the entire codebase yet, so maybe there is a way.

jglick commented on core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java:
```diff
@@ -70,16 +70,20 @@ public boolean prebuild(AbstractBuild<?,?> build, BuildListener listener) {
             // delegate to the overloaded version defined in SimpleBuildStep
             FilePath workspace = build.getWorkspace();
-            if (workspace == null) {
+            if (workspace == null && sbs.requiresWorkspace()) {
```
Note that this would be a pathological condition: the build is almost certainly broken anyway.

jglick commented on core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java:
```diff
@@ -70,16 +70,20 @@ public boolean prebuild(AbstractBuild<?,?> build, BuildListener listener) {
             }
-            ((SimpleBuildStep) this).perform(build, workspace, build.getEnvironment(listener), launcher, listener);
+            if (launcher == null && sbs.requiresLauncher()) {
+                throw new AbortException("no launcher for " + build);
```
Similarly, for an `AbstractBuild` this would be a pathological condition. There is probably no real need to patch this class at all.

Zastai replied:
File changes reverted.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -77,15 +78,15 @@
      */
     @Deprecated
-    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
+    default void perform(@NonNull Run<?, ?> run, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher,
```
```suggestion
    default void perform(@NonNull Run<?, ?> run, @Nullable FilePath workspace, @Nullable Launcher launcher,
```

since whether or not null is actually allowed is dependent on complex things the static analyzer will not understand

Zastai replied:
It was not particularly clear to me when each of those two was applicable; I rather assumed they were equivalent so I just used the same one everywhere. Will have a look around and adjust these.

jglick replied:
> The annotated element could be null under some circumstances.
> In general, this means developers will have to read the documentation to determine when a null value is acceptable and whether it is necessary to check for a null value.

Zastai replied:
Switched `@CheckForNull` to `@Nullable` as suggested

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -94,15 +95,15 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
      * @since TODO
      */
-    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher,
+    default void perform(@NonNull Run<?, ?> run, @CheckForNull FilePath workspace, @NonNull EnvVars env, @CheckForNull Launcher launcher,
```
```suggestion
    default void perform(@NonNull Run<?, ?> run, @Nullable FilePath workspace, @NonNull EnvVars env, @Nullable Launcher launcher,
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -68,14 +68,40 @@
      * @throws InterruptedException if setup is interrupted
      */
-    public abstract void setUp(Context context, Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException;
+    public abstract void setUp(@NonNull Context context, @NonNull Run<?,?> build, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher, @NonNull TaskListener listener, @NonNull EnvVars initialEnvironment) throws IOException, InterruptedException;
```
```suggestion
    public abstract void setUp(@NonNull Context context, @NonNull Run<?,?> build, @Nullable FilePath workspace, @Nullable Launcher launcher, @NonNull TaskListener listener, @NonNull EnvVars initialEnvironment) throws IOException, InterruptedException;
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -120,13 +146,39 @@ public void setDisposer(@NonNull Disposer disposer) {
          * @throws InterruptedException if tear down is interrupted
          */
-        public abstract void tearDown(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException;
+        public abstract void tearDown(@NonNull Run<?,?> build, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher, @NonNull TaskListener listener) throws IOException, InterruptedException;
```
```suggestion
        public abstract void tearDown(@NonNull Run<?,?> build, @Nullable FilePath workspace, @Nullable Launcher launcher, @NonNull TaskListener listener) throws IOException, InterruptedException;
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -120,13 +146,39 @@ public void setDisposer(@NonNull Disposer disposer) {
-        public abstract void tearDown(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException;
+        public abstract void tearDown(@NonNull Run<?,?> build, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher, @NonNull TaskListener listener) throws IOException, InterruptedException;
+
+        /**
+         * Indicates whether or not this wrapper callback requires a launcher.
+         * <p>
+         * If this return {@code false}, this means that {@link #tearDown(Run, FilePath, Launcher, TaskListener)}
+         * will accept {@code null} for its {@code launcher} parameter.
+         *
+         * @return {@code true} when this wrapper callback requires a launcher; {@code false} otherwise.
+         * @since TODO
+         */
+        public boolean requiresLauncher() {
+            return true;
+        }
+
+        /**
+         * Indicates whether or not this wrapper callback requires a workspace.
+         * <p>
+         * If this return {@code false}, this means that {@link #tearDown(Run, FilePath, Launcher, TaskListener)}
+         * will accept {@code null} for its {@code workspace} parameter.
+         *
+         * @return {@code true} when this wrapper callback requires a workspace; {@code false} otherwise.
+         * @since TODO
+         */
+        public boolean requiresWorkspace() {
+            return true;
+        }
```
Redundant. Either the wrapper setup & teardown both require a node context (~ workspace + launcher), or neither do.

Zastai replied:
The `Disposer` class is public API, not restricted to `SimpleBuildWrapper`s; so I figured it made sense to give it support for separate configuration.

Also, without these methods, there is nothing to refer to in the tearDown() JavaDoc to explain when the argument may be null, because there is no link to the original wrapper in this class.

An org search suggests there is no plugin using `Disposer` outside a wrapper context, so if you're happy with just making the workspace/launcher nullable, then I can indeed drop these.

jglick replied:
Yes a disposer matches a wrapper. It is only a separate type for reasons of serialization: you must carefully manage state of the disposer.

Zastai replied:
Methods dropped; JavaDoc adjusted.

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -138,7 +190,7 @@ protected boolean runPreCheckout() {
     }
 
-    @Override public final Environment setUp(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+    @Override public final Environment setUp(@NonNull AbstractBuild build, final @CheckForNull Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```
```suggestion
    @Override public final Environment setUp(@NonNull AbstractBuild build, final @Nullable Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -148,7 +200,7 @@ protected boolean runPreCheckout() {
     }
 
-    @Override public final void preCheckout(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+    @Override public final void preCheckout(@NonNull AbstractBuild build, final @CheckForNull Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```
```suggestion
    @Override public final void preCheckout(@NonNull AbstractBuild build, final @Nullable Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -157,9 +209,9 @@ protected boolean runPreCheckout() {
-        private final Launcher launcher;
-        EnvironmentWrapper(Context c, Launcher launcher) {
+        private final @NonNull Context c;
+        private final @CheckForNull Launcher launcher;
```
```suggestion
        private final @Nullable Launcher launcher;
```

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -157,9 +209,9 @@ protected boolean runPreCheckout() {
-        EnvironmentWrapper(Context c, Launcher launcher) {
+        private final @NonNull Context c;
+        private final @CheckForNull Launcher launcher;
+        EnvironmentWrapper(@NonNull Context c, @CheckForNull Launcher launcher) {
```
```suggestion
        EnvironmentWrapper(@NonNull Context c, @Nullable Launcher launcher) {
```

jglick commented:
https://github.com/jenkinsci/jenkins/pull/4784#discussion_r445877019 should be addressed, I think: the API is unnecessarily broad. Everything else looks fine.

Zastai committed:
Use @Nullable instead of @CheckForNull.

Zastai committed:
Remove changes to BuildStepCompatibilityLayer.

Deemed unnecessary because AbstractBuild always has a
workspace/launcher context.

Zastai committed:
Remove more changes to BuildStepCompatibilityLayer.

Zastai committed:
Simplify SimpleBuildWrapper.Disposer.

Zastai committed:
Merge branch 'master' into JENKINS-46175

jglick commented on core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:
```diff
@@ -120,13 +147,14 @@ public void setDisposer(@NonNull Disposer disposer) {
          * @param build a build being run
-         * @param workspace a workspace of the build
-         * @param launcher a way to start commands
+         * @param workspace a workspace of the build; may be {@code null} if is allowed by the corresponding wrapper
+         * @param launcher a way to start commands; may be {@code null} if is allowed by the corresponding wrapper
```
```suggestion
         * @param workspace a workspace of the build; may be {@code null} if allowed by the corresponding wrapper
         * @param launcher a way to start commands; may be {@code null} if allowed by the corresponding wrapper
```

Zastai replied:
Yes, I seem to have either lost an `it` or failed to remove the `is`.

Zastai committed:
Fix English in doc comment

Zastai committed:
Merge branch 'master' into JENKINS-46175

oleg-nenashev reviewed (commented):
I am not a big fan of using `Nullable` in API, because it does not really enforce any checks in the static analysis tools. TBH My preference would be to introduce new API endpoints or new child classes for workspace-less operations, but it could also overcomplicate APi as well. I am not bloking this pull request though

Zastai commented:
I'll look at drafting an alternative PR that adds extra API instead.

jglick commented:
`@Nullable` is not ideal, but on balance I think this is preferable to #4829. No strong opinion.

oleg-nenashev reviewed (approved):
I prefer #4829, but this one will do the job as well. I approve this pull request to indicate that I can live with that if other core maintainers prefer this one

MRamonLeon reviewed (commented):
I would like to see some checks for the `Nullable` parameters. If they're null, then the `requiresXXX` method has to return true. Other way, it's up to the developer. And we don't trust developers :wink: 

I like more #4829 for this reason.

Zastai commented:
Closing - this has been replaced by https://github.com/jenkinsci/jenkins/pull/4829

reinholdfuereder commented on core/src/main/java/jenkins/tasks/SimpleBuildStep.java:
```diff
@@ -115,6 +116,32 @@ default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNu
+     * will accept {@code null} for its {@code workspace} parameter.
+     *
+     * @return {@code true} when this step requires a workspace; {@code false} otherwise.
+     * @since TODO
```
It looks like there still 3 little open TODOs in this PR, cf. "@since TODO"

Zastai replied:
Aside from this PR being closed, those are to be the version number of the first release that includes the API. Because this is a moving target, it is usually left to be filled in after the PR is merged.
